app_dart/lib/src/request_handlers/push_build_status_to_github.dart

```diff
@@ -28,79 +30,95 @@ class PushBuildStatusToGithub extends ApiRequestHandler<Body> {
-          update.status = buildStatus.githubStatus;
-          update.updates += 1;
-          updates.add(update);
-        } catch (error) {
-          log.error(
-              'Failed to post status update to ${slug.fullName}#${pr.number}: $error');
+    final List<Branch> branches = await getBranches(
+        config, branchHttpClientProvider, log, gitHubBackoffCalculator);
+    for (Branch branch in branches) {
+      final BuildStatus buildStatus = await buildStatusProvider
+          .calculateCumulativeStatus(branch: branch.name);
+      final GitHub github = await config.createGitHubClient();
+      final List<GithubBuildStatusUpdate> updates = <GithubBuildStatusUpdate>[];
+      log.debug('Computed build result of $buildStatus');
+
+      // Insert build status to bigquery.
+      await _insertBigquery(buildStatus, branch.name);
+
+      final List<PullRequest> pullRequests =
+          await githubService.listPullRequests(slug, branch.name);
+      for (PullRequest pr in pullRequests) {
+        final GithubBuildStatusUpdate update =
+            await datastore.queryLastStatusUpdate(slug, pr);
+
+        if (update.status != buildStatus.githubStatus) {
+          log.debug(
+              'Updating status of ${slug.fullName}#${pr.number} from ${update.status}');
+          final CreateStatus request = CreateStatus(buildStatus.githubStatus);
+          request.targetUrl =
+              'https://flutter-dashboard.appspot.com/build.html';
+          request.context = 'flutter-build';
+          if (buildStatus != BuildStatus.succeeded) {
+            request.description =
+                'Flutter build is currently broken. Please do not merge this '
+                'PR unless it contains a fix to the broken build.';
+          }
+
+          try {
+            await github.repositories.createStatus(slug, pr.head.sha, request);
+            update.status = buildStatus.githubStatus;
+            update.updates += 1;
+            updates.add(update);
+          } catch (error) {
+            log.error(
+                'Failed to post status update to ${slug.fullName}#${pr.number}: $error');
+          }
         }
       }
-    }
 
-    final int maxEntityGroups = config.maxEntityGroups;
-    for (int i = 0; i < updates.length; i += maxEntityGroups) {
-      await datastore.db.withTransaction<void>((Transaction transaction) async {
-        transaction.queueMutations(
-            inserts: updates.skip(i).take(maxEntityGroups).toList());
-        await transaction.commit();
-      });
+      final int maxEntityGroups = config.maxEntityGroups;
+      for (int i = 0; i < updates.length; i += maxEntityGroups) {
+        await datastore.db
+            .withTransaction<void>((Transaction transaction) async {
+          transaction.queueMutations(
+              inserts: updates.skip(i).take(maxEntityGroups).toList());
+          await transaction.commit();
+        });
+      }
+      log.debug('Committed all updates');
```

Comment:
This may help you understand: https://github.com/flutter/cocoon/blob/master/app_dart/lib/src/datastore/cocoon_config.dart#L61

Reply:
With such a nitty detail to datastore, it would be cleaner to move this code to the Datastore service and add documentation to why it's needed. Copying that explanation is a good start, but i'm still confused as to why pushing build status to github can handle Datastore writes

Reply:
We have a status entity in datastore to denote how many updates we have made so for for existing PRs. So whenever we push status to github, we write corresponding update index.

Reply:
Sounds like it would be good to document it in the code then :)

Reply:
Added a brief documentation. Thank you!